esp-hal/src/work_queue.rs

```diff
@@ -0,0 +1,451 @@
+//! post work items into the work queue, and poll for completion. The act of polling
+//! processes the queue, which allows any number of clients to post work without deadlocking.
+//!
+//! Work queues are configured by backends. Backends register a `process` callback which is
+//! called when a client posts a work item or polls for completion.
+//!
+//! Posting a work item into the queue returns a handle. The handle can be used to poll whether
+//! the work item has been processed. Dropping the handle will cancel the work item.
+#![cfg_attr(esp32c2, allow(unused))]
+
+use core::{future::poll_fn, marker::PhantomData};
+
+use crate::{asynch::AtomicWaker, sync::Locked};
+
+pub(crate) struct VTable<T: Sync> {
+    /// Starts processing a new work item.
+    ///
+    /// The function returns whether the work item was accepted. If there is no driver currently
+    /// processing the queue, this function will return false to prevent removing the work item
+    /// from the queue.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) post: fn(*const (), &mut T) -> bool,
+
+    /// Polls the status of a particular work item.
+    ///
+    /// Return `None` if the work item is not currently being processed.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) poll: fn(*const (), &mut T) -> Option<Poll>,
+
+    /// Attempts to abort processing a work item.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) cancel: fn(*const (), &mut T),
+
+    /// Called when the work queue becomes empty.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) on_empty: fn(*const ()),
+}
+
+impl<T: Sync> VTable<T> {
+    pub(crate) const fn noop() -> Self {
+        Self {
+            post: |_, _| false,
+            poll: |_, _| None,
+            cancel: |_, _| (),
+            on_empty: |_| (),
+        }
+    }
+}
+
+struct Inner<T: Sync> {
+    head: *mut WorkItem<T>,
+    tail: *mut WorkItem<T>,
+    current: *mut WorkItem<T>,
+
+    data: *const (),
+    vtable: VTable<T>,
+}
+
+impl<T: Sync> Inner<T> {
+    /// Places a work item at the end of the queue.
+    fn enqueue(&mut self, ptr: *mut WorkItem<T>) {
+        if self.tail.is_null() {
+            // Queue is empty, set both `head` and `tail`.
+            self.tail = ptr;
+            self.head = ptr;
+        } else {
+            unsafe {
+                // Safety: we just checked that `tail` is not null.
+                (*self.tail).next = ptr;
+            }
+        }
+    }
+
+    /// Places a work item at the front of the queue.
+    fn enqueue_front(&mut self, ptr: *mut WorkItem<T>) {
+        unsafe { (*ptr).next = self.head };
+        self.head = ptr;
+        if self.tail.is_null() {
+            self.tail = ptr;
+        }
+    }
+
+    /// Pops and returns a work item from the start of the queue.
+    fn dequeue(&mut self) -> Option<*mut WorkItem<T>> {
```

Comment:
The return type here is a bit weird, might be worth considering `NonNull` instead of `*mut`